boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,293 @@
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        max_angle_range=(45, -45),
+    ):
+        """Initializes the class attributes.
+
+        Args:
+            `current_heading` (Scalar): Current boat heading direction in degrees.
+            `desired_heading` (Scalar): Target boating heading direction in degrees.
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Timeseries of error values computed over time.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range
+
+        """
+        super().__init__(current_heading, desired_heading, current_control_ang, time_step)
+
+        self.kp = kp
+        self.cp = cp
+        self.max_angle_range = max_angle_range  # passed in degrees
+        self.running_error = 0
+
+    def compute_error(
+        self, desired: Scalar, current: Scalar
+    ) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+
+        Args:
+            `desired` (Scalar): New desired heading
+            `current` (Scalar): New current heading
+
+        Returns:
+            Scalar: The error between the given headings in radians.
+        """
+        current_bound = bound_to_180(current)
+        desired_bound = bound_to_180(desired)
+        desired_rad = np.deg2rad(desired)
+        current_rad = np.deg2rad(current)
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        if current_bound == desired_bound:
+            error = 0
+
+        return error
+
+    def compute_feedback_angle(self) -> Scalar:
+        """Computes the feedback angle from the given error (in radians)
+        Calculations done using Raye's implementation
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Difference between current and target rudder angle in degrees.
+        """
+        error = self.compute_error(self.current_heading, self.desired_heading)
+
+        if abs(error) > pi:
+            raise ValueError("heading_error must be between -pi and pi")
+
+        rudder_change = (self.kp * error) / (1 + (self.cp * error))
+        self.running_error = np.rad2deg(rudder_change)
+
+        return np.rad2deg(rudder_change)
+
+    def compute_setpoint(self) -> Scalar:
+        """Calculates the rudder setpoint angle. Only called at the very beginning once
+        to determine the setpoint angle
+
+        Returns:
+            The rudder setpoint angle in degrees
+        """
+        return self.prev_control[0] + self.compute_feedback_angle()
```

Comment:
Is this current angle + change?

Reply:
That was brainfart I originally had it as self.prev_control[-1] but I was like oh this should use the original starting rudder_control_angle. Using index -1 should call the current I think.